### PR TITLE
Tools:  Allow using Optional<T> method parameters

### PR DESCRIPTION
**Pull request by zhangdashuai11**
when use tools  if user define method parameter as Optional addOptionalParameter

### CODE REVIEW

**Comment by langchain4j** [comment]
@zhangdashuai11 please add tests and document this change. Please see https://github.com/langchain4j/langchain4j/blob/main/CONTRIBUTING.md

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 1]
```diff
@@ -56,6 +57,10 @@ public static ToolSpecification toolSpecificationFrom(Method method) {
             if (parameter.isAnnotationPresent(ToolMemoryId.class)) {
                 continue;
             }
+            //if user define method parameter as Optional addOptionalParameter
+            if (parameter.getType().equals(Optional.class)) {
+                builder.addOptionalParameter(parameter.getName(), toJsonSchemaProperties(parameter));
+            }
             builder.addParameter(parameter.getName(), toJsonSchemaProperties(parameter));
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:64]: This should probybly be in `else` block.

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java [hunk 2]
```diff
@@ -56,6 +57,10 @@ public static ToolSpecification toolSpecificationFrom(Method method) {
             if (parameter.isAnnotationPresent(ToolMemoryId.class)) {
                 continue;
             }
+            //if user define method parameter as Optional addOptionalParameter
+            if (parameter.getType().equals(Optional.class)) {
+                builder.addOptionalParameter(parameter.getName(), toJsonSchemaProperties(parameter));
```
Comment (langchain4j) [inline langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: `toJsonSchemaProperties` will receive `Optional` type, but should probably receive an actual type

**Review by ghost — CHANGES_REQUESTED** [review]

File: langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java (hunk 2 above)
Comment (Kugaaa) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: `dev.langchain4j.agent.tool.DefaultToolExecutor#coerceArgument` also need packe `Optional`?

Also, I have a question: Is `Optional` suitable as a parameter?
[Why should Java 8's Optional not be used in arguments](https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments)

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: I have no strong opinion on this, as always it depends...

```java
@Tool
void method(@P(optional = true) String parameter) {...}
```

or 

```java
@Tool
void method(Optional<String> parameter) {...}
```

what would you choose?

Comment (Kugaaa) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: I vote for option one~

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: @geoand @jmartisk WDYT?

Comment (geoand) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: I don't have a super strong opinion either, but I propably slightly lean towards `1`

Comment (langchain4j) [reply langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java:62]: Ok, I guess it does not harm to have both options, but for now let's go with the first one then

**Comment by langchain4j** [comment]
Implemented in https://github.com/langchain4j/langchain4j/pull/1370

**Pull request closed without merging**